### PR TITLE
[UF-403] Fork and Mirror

### PR DESCRIPTION
**Pull request by adrielparedes**
The goal is to have the possibility to fork any git repository inside the niogit directory.
To complete the goal we need:
* Separate git repositories into different directories. Those directories are named like the Organizational Units.
* Create a Fork command that copy all the content of the repository to fork and add the reference to the internal path of the origin repository.
* Write information about fork so we can track where those forks are, how many forks has those repositories, etc.

### CODE REVIEW

File: uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-jgit/src/main/java/org/uberfire/java/nio/fs/jgit/util/commands/Fork.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+            if ( logger.isDebugEnabled() ) {
+                logger.debug( "Repository <{}> forked successfuly from <{}>", target, source );
+            }
+
+        } catch ( IOException | GitAPIException e ) {
+            throw new GitException( "Cannot fork repository", e );
+        }
+
+        if ( gitDestination != null ) {
```
Comment (porcelli) [inline uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-jgit/src/main/java/org/uberfire/java/nio/fs/jgit/util/commands/Fork.java]: return Optional.ofNullable( gitDestination )

File: uberfire-backend/uberfire-backend-server/src/test/java/org/uberfire/backend/server/io/object/ObjectStorageTest.java [hunk 2]
```diff
@@ -127,6 +127,16 @@ public void testWriteTwoTimes() {
     }
 
     @Test
+    public void testRenameFile() {
```
Comment (ederign) [inline uberfire-backend/uberfire-backend-server/src/test/java/org/uberfire/backend/server/io/object/ObjectStorageTest.java]: move?

**Comment by manstis** [comment]
Please note this PR *appears* to cause test failures in ```drools-wb``` (and I suspect ``kie-wb-common```). This was evidenced by (the now closed) [PR](https://github.com/droolsjbpm/drools-wb/pull/288). Can I please request you test compile (with tests) at least ```kie-wb-common``` and ```drools-wb``` - although other repositories containing "indexing tests" may equally be affected (e.g. ```jbpm-designer``` and ```optaplanner-wb```). Thank-you.

File: uberfire-backend/uberfire-backend-server/src/test/java/org/uberfire/backend/server/io/object/ObjectStorageTest.java (hunk 2 above)
Comment (adrielparedes) [inline uberfire-backend/uberfire-backend-server/src/test/java/org/uberfire/backend/server/io/object/ObjectStorageTest.java]: @ederign yes... you think is useless?

**Comment by adrielparedes** [comment]
@manstis sure, I'm going to check what happened and I will let you know

File: uberfire-api/src/main/java/org/uberfire/backend/vfs/PathFactory.java [hunk 3]
```diff
@@ -87,10 +87,10 @@ public static Path newLockPath( final Path path) {
      * Examples:
      * 
      * <pre>
-     * default://locks@system/repo/master/some/path/to/file.txt.ulock =>
+     * default://locks@system/system/repo/master/some/path/to/file.txt.ulock =>
      * default://master@repo/some/path/to/file.txt
      * 
-     * file:\\locks@system\repo\master\some\path\to\file.txt.ulock =>
+     * file:\\locks@system/system\repo\master\some\path\to\file.txt.ulock =>
```
Comment (manstis) [inline uberfire-api/src/main/java/org/uberfire/backend/vfs/PathFactory.java]: Should this be a backslash (```\\locks@system\system\repo....```)?

File: uberfire-api/src/main/java/org/uberfire/backend/vfs/PathFactory.java [hunk 4]
```diff
@@ -61,10 +61,10 @@ public static Path newLock( final Path path ) {
      * 
      * <pre>
      * default://master@repo/some/path/to/file.txt =>
-     * default://locks@system/repo/master/some/path/to/file.txt.ulock
+     * default://locks@system/system/repo/master/some/path/to/file.txt.ulock
      * 
      * file:\\master@repo\some\path\to\file.txt =>
-     * file:\\locks@system\repo\master\some\path\to\file.txt.ulock
+     * file:\\locks@system/system\repo\master\some\path\to\file.txt.ulock
```
Comment (manstis) [inline uberfire-api/src/main/java/org/uberfire/backend/vfs/PathFactory.java]: Should this be a backslash (\\locks@system\system\repo....)?

**Comment by adrielparedes** [comment]
@manstis please, check if compilations worked, I fixed some stuffs. Thanks!

**Comment by ederign** [comment]
@adrielparedes I'll be away for 1 week. Meanwhile, talk with @csadilek about this PR!

**Comment by adrielparedes** [comment]
@ederign np, I will talk with him. Have fun!

**Comment by adrielparedes** [comment]
I have revert some changes to have the original behavior but adding the possibility to migrate a repository. To do the migration you have to declare a environment variable at the creation of a new file system. In that variable you set the URI of the repository you want to migrate.

If the old repository exists and the new repository doesn't, it will migrate the old repo to the new one.
If the old repository doesn't exists, it will just create the new one.
And if the new repository exists it will not do any migration.

----
*PluginsFileSystemBuilder.java*
```
public static final String OLD_SYSTEM_PLUGINS = "default://plugins";
public static final String DEFAULT_SYSTEM_PLUGINS = "default://system/plugins";
...
try {
            return ioService.newFileSystem( URI.create( DEFAULT_SYSTEM_PLUGINS ),
                                            new HashMap<String, Object>() {{
                                                put( "init", Boolean.TRUE );
                                                put( "internal", Boolean.TRUE );
                                                put( "migrate-from", URI.create( OLD_SYSTEM_PLUGINS ) );
                                            }} );
        } catch ( final FileSystemAlreadyExistsException e ) {
            return ioService.getFileSystem( URI.create( DEFAULT_SYSTEM_PLUGINS ) );
        }
```

@porcelli @ederign @csadilek

**Comment by csadilek** [comment]
+1, if not already done, the WB needs to be smoke tested with these changes before the merge. Thanks.

**Comment by manstis** [comment]
@porcelli @adrielparedes @csadilek I suspect this change has broken some things: https://kie-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/kie-wb-distributions/1403/testReport/ Since this PR was merged there are 22 failing integration tests relating to ```null``` "origin" (in presumably the ```Repository``` class). Could you please investigate?

**Comment by manstis** [comment]
@porcelli @adrielparedes @csadilek I am also unable to clone any repository due to the following error:

```
java.lang.IllegalArgumentException: Parameter named 'origin' should be not null!
	at org.uberfire.commons.validation.PortablePreconditions.checkNotNull(PortablePreconditions.java:119)
	at org.guvnor.structure.repositories.impl.metadata.GitMetadataImpl.<init>(GitMetadataImpl.java:53)
	at org.guvnor.structure.repositories.impl.metadata.GitMetadataImpl.<init>(GitMetadataImpl.java:41)
	at org.guvnor.structure.backend.repositories.git.GitMetadataStoreImpl.write(GitMetadataStoreImpl.java:63)
	at org.guvnor.structure.backend.repositories.RepositoryServiceImpl.createRepository(RepositoryServiceImpl.java:229)
```

**Comment by adrielparedes** [comment]
@manstis I found the error, let me create a new PR with the fix after I do the correct smoke test

**Pull request merged**